fplll/hnf.cpp

```diff
@@ -0,0 +1,346 @@
+  int l = (c - r) * (c > r);
+
+  /* main loop, decreases from last column to the limit */
+  for (j = c - 1, k = r - 1; j >= l; j--, k--)
+  {
+    // checks if everything above diagonal zero
+    for (i = k - 1; i >= 0; i--)
+    {
+      if (B[i][j] != 0)
+      {
+        return 1;
+      }
+    }
+    // checks if diagonal is non-negative
+    if (B[k][j] < 0)
+    {
+      return 1;
+    }
+    // if "diagonal" entry is zero, we have to "shift" the diagonal position
+    else if (B[k][j] == 0)
+    {
+      // pivot row position doesn't change, increment to compensate
+      k++;
+      // lower the limit as we skipped one column without increasing row
+      if (l > 0)
+      {
+        l--;
+      }
+    }
+    else
+    {
+      // checks if everything below is positive and below the diagonal
+      for (i = k + 1; i < r; i++)
+      {
+        if ((B[i][j] > B[k][j]) || (B[i][j] < 0))
+        {
+          return 1;
+        }
+      }
+    }
+  }
+
+  return 0;
+}
+
+int in_hnf(ZZ_mat<mpz_t> &B, const vector<Z_NR<mpz_t>> &w)
+{
+
+  vector<Z_NR<mpz_t>> v = w;
+  Z_NR<mpz_t> q;
+
+  /* matrix bounds */
+  int r = B.get_rows(), c = B.get_cols();
+  /* test if the vector size is correct */
+  if ((int)v.size() != c)
+  {
+    cerr << "in_hnf error : matrix-vector sizes do not match\n";
+    return -1;
+  }
+  /* matrix indexes (k = pivot)*/
+  int i, j, j2, k;
+  /* the limit "l" depends of matrix dimensions */
+  int l = (c - r) * (c > r);
+
+  /* main loop, decreases from last column to the limit */
+  for (j = c - 1, k = r - 1; j >= l; j--, k--)
+  {
+    // checks if everything above diagonal zero
+    for (i = k - 1; i >= 0; i--)
+    {
+      if (B[i][j] != 0)
+      {
+        return 1;
+      }
+    }
+    // checks if diagonal is non-negative
+    if (B[k][j] < 0)
+    {
+      return 1;
+    }
+    // if "diagonal" entry is zero, we have to "shift" the diagonal position
+    else if (B[k][j] == 0)
+    {
+      // pivot row position doesn't change, increment to compensate
+      k++;
+      // lower the limit as we skipped one column without increasing row
+      if (l > 0)
+      {
+        l--;
+      }
+    }
+    else
+    {
+      // checks if everything below is positive and below the diagonal
+      for (i = k + 1; i < r; i++)
+      {
+        if ((B[i][j] > B[k][j]) || (B[i][j] < 0))
+        {
+          return 1;
+        }
+      }
+      // reduces the vector by the row pivot vector for membership test
+      q.fdiv_q(v[j], B[k][j]);
+      for (j2 = j; j2 >= 0; j2--)
+      {
+        v[j2].submul(q, B[k][j2]);
+      }
+    }
+  }
+
+  // membership test : the vector after reduction should be a zero one.
+  for (j = 0; j < c; j++)
+  {
+    if (v[j] != 0)
+    {
+      return -1;
+    }
+  }
+
+  return 0;
+}
+
+int in_hnf(ZZ_mat<mpz_t> &B, const ZZ_mat<mpz_t> &A)
+{
+
+  ZZ_mat<mpz_t> tmp = A;
+  Z_NR<mpz_t> q;
+
+  /* matrix bounds */
+  int r = B.get_rows(), c = B.get_cols();
```

Comment:
This looks like copy'n'pasted code again or am I missing something?

Reply:
It actually look both like copy-pasted code because it actually is for most of it. I didn't want to make a second loop for the reduction part that detects at the same time if there's a break on the diagonal form (not all HNF have a strictly diagonal form, it also depends of rank and sizes), so I included the reduction part in the detection all together so I can have only a single loop and not have another diagonal pivot detection to run when reducing the copy of the vector/matrix for checking if it is in the lattice.

Reply:
Maybe the common code could be factored out into a separate helper function which is then called by the higher level functions?

Reply:
I'll try that.